File: 105. Construct Binary Tree from Preorder and Inorder Traversal/105. Construct Binary Tree from Preorder and Inorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private Map<Integer, Integer> inMap;
    private int[] preorder;
    private int[] inorder;
    
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        this.preorder = preorder;
        this.inorder = inorder;
        inMap = new HashMap<>();
        
        for (int i = 0; i < inorder.length; i++) {
            inMap.put(inorder[i], i);
        }
        
        return buildTreeHelper(0, preorder.length - 1, 0, inorder.length - 1);
    }
    
    private TreeNode buildTreeHelper(int preStart, int preEnd, int inStart, int inEnd) {
        if (preStart > preEnd) return null;
        
        TreeNode root = new TreeNode(preorder[preStart]);
        int rootIndex = inMap.get(root.val);
        int leftSubtreeSize = rootIndex - inStart;
        
        root.left = buildTreeHelper(preStart + 1, preStart + leftSubtreeSize,
                                  inStart, rootIndex - 1);
        root.right = buildTreeHelper(preStart + leftSubtreeSize + 1, preEnd,
                                   rootIndex + 1, inEnd);
        
        return root;
    }
}

File: 105. Construct Binary Tree from Preorder and Inorder Traversal/105. Construct Binary Tree from Preorder and Inorder Traversal.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::HashMap;

impl Solution {
    fn build_tree_helper(
        preorder: &[i32],
        pre_start: usize,
        pre_end: usize,
        inorder: &[i32],
        in_start: usize,
        in_end: usize,
        in_map: &HashMap<i32, usize>
    ) -> Option<Rc<RefCell<TreeNode>>> {
        if pre_start > pre_end {
            return None;
        }
        
        let mut root = TreeNode::new(preorder[pre_start]);
        let root_index = in_map[&preorder[pre_start]];
        let left_subtree_size = root_index - in_start;
        
        root.left = Self::build_tree_helper(
            preorder,
            pre_start + 1,
            pre_start + left_subtree_size,
            inorder,
            in_start,
            root_index - 1,
            in_map
        );
        
        root.right = Self::build_tree_helper(
            preorder,
            pre_start + left_subtree_size + 1,
            pre_end,
            inorder,
            root_index + 1,
            in_end,
            in_map
        );
        
        Some(Rc::new(RefCell::new(root)))
    }
    
    pub fn build_tree(preorder: Vec<i32>, inorder: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        let mut in_map = HashMap::new();
        for (i, &val) in inorder.iter().enumerate() {
            in_map.insert(val, i);
        }
        
        Self::build_tree_helper(
            &preorder,
            0,
            preorder.len() - 1,
            &inorder,
            0,
            inorder.len() - 1,
            &in_map
        )
    }
}